plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilter.java

```diff
@@ -137,4 +151,91 @@ public void failed(Exception e) {
     public int order() {
         return Integer.MAX_VALUE;
     }
+
+    final static class BulkRequestSource implements Iterator<ActionRequest> {
+
+        final BulkRequest bulkRequest;
+        final ActionListener<BulkResponse> actionListener;
+
+        final Set<Integer> failedSlots;
+        final List<BulkItemResponse> itemResponses;
+
+        int currentSlot = -1;
+
+        BulkRequestSource(BulkRequest bulkRequest, ActionListener<BulkResponse> actionListener) {
+            this.bulkRequest = bulkRequest;
+            this.actionListener = actionListener;
+            this.failedSlots = new HashSet<>();
+            this.itemResponses = new ArrayList<>(bulkRequest.requests().size());
+        }
+
+        @Override
+        public ActionRequest next() {
+            return bulkRequest.requests().get(++currentSlot);
+        }
+
+        @Override
+        public boolean hasNext() {
+            return (currentSlot + 1) < bulkRequest.requests().size();
+        }
+
+        Tuple<BulkRequest, ActionListener<BulkResponse>> bulkRequestAndActionListener() {
+            if (itemResponses.isEmpty()) {
+                return new Tuple<>(bulkRequest, actionListener);
+            } else {
+                BulkRequest modifiedBulkRequest = new BulkRequest(bulkRequest);
+                modifiedBulkRequest.refresh(bulkRequest.refresh());
+                modifiedBulkRequest.consistencyLevel(bulkRequest.consistencyLevel());
+                modifiedBulkRequest.timeout(bulkRequest.timeout());
+
+                List<Integer> originalSlots = new ArrayList<>();
+                for (int i = 0; i < bulkRequest.requests().size(); i++) {
+                    ActionRequest request = bulkRequest.requests().get(i);
+                    if (failedSlots.contains(i) == false) {
+                        originalSlots.add(i);
+                        modifiedBulkRequest.add(request);
+                    }
+                }
+                ActionListener<BulkResponse> actionListener = new IngestBulkResponseListener(originalSlots, itemResponses, this.actionListener);
+                return new Tuple<>(modifiedBulkRequest, actionListener);
+            }
+        }
+
+        void markCurrentItemAsFailed(IndexRequest indexRequest, Throwable e) {
```

Comment:
do we need to pass in the index request here? isn't saying "current" enough to retrieve the current index request from the original bulk request?